New module: SoupX ambient RNA removal

LeonHafner opened the pull request:
## PR checklist

- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`

LeonHafner committed:
Added soupx module

LeonHafner commented:
@nf-core-bot fix linting

LeonHafner committed:
add full meta.yml

LeonHafner committed:
fix mixedCase linting

LeonHafner committed:
fix singularity container tmp issue

LeonHafner committed:
fix linting

LeonHafner committed:
Merge branch 'master' into soupx

LeonHafner committed:
Change hashes to match CI

nictru commented:
Should switch this to an R implementation. [This Vignette](https://cellgeni.github.io/notebooks/html/new-10kPBMC-SoupX.html) can serve as a template. Should use AnndataR if the file ends with `.h5ad` otherwise `readRDS`.

LeonHafner commented:
@nictru what's the plan with this PR? AnnDataR isn’t available on bioconda yet, as it hasn’t had its first release.

famosab commented:
@LeonHafner @nictru Are you still working on this PR?

LeonHafner commented:
New information about `anndataR` and when it will be added to bioconda: https://github.com/scverse/anndataR/issues/254#issuecomment-2731847030

nictru committed:
Merge branch 'master' into soupx

SPPearce commented:
@LeonHafner , @nictru , any progress?

LeonHafner committed:
Merge branch 'master' into soupx

LeonHafner committed:
Merge branch 'nf-core:master' into soupx

LeonHafner committed:
switch to python

LeonHafner committed:
add topics

LeonHafner committed:
Merge branch 'master' into soupx

LeonHafner committed:
fix env.yml

LeonHafner committed:
fix linting

SPPearce commented on modules/nf-core/soupx/tests/main.nf.test:
```diff
@@ -0,0 +1,59 @@
+            assertAll(
+                { assert process.success },
+                { assert snapshot(process.out).match() }
+            )
```
```suggestion
            assert process.success
            assert snapshot(sanitizeOutput(process.out)).match()
```

SPPearce commented on modules/nf-core/soupx/tests/main.nf.test:
```diff
@@ -0,0 +1,59 @@
+            assertAll(
+                { assert process.success },
+                { assert snapshot(process.out).match() }
+            )
```
```suggestion
            assert process.success
            assert snapshot(sanitizeOutput(process.out)).match()
```

SPPearce commented on modules/nf-core/soupx/main.nf:
```diff
@@ -0,0 +1,38 @@
+    task.ext.when == null || task.ext.when
+
+    script:
+    prefix = task.ext.prefix ?: "${meta.id}"
```
Set something as a default prefix so it doesn't automatically clash.
```suggestion
    prefix = task.ext.prefix ?: "${meta.id}_soupx"
```

SPPearce commented on modules/nf-core/soupx/main.nf:
```diff
@@ -0,0 +1,38 @@
+    template 'soupx.R'
+
+    stub:
+    prefix = task.ext.prefix ?: "${meta.id}"
```
```suggestion
    prefix = task.ext.prefix ?: "${meta.id}_soupx"
```

SPPearce commented on modules/nf-core/soupx/main.nf:
```diff
@@ -0,0 +1,38 @@
+    tuple val(meta), path(h5ad_filtered), path(h5ad_raw)
+
+    output:
+    tuple val(meta), path("*.h5ad"), emit: h5ad
```
Please add the disambiguation check to ensure that the name of the h5ad is not the same as either input.

SPPearce reviewed (commented):
There are a number of hardcoded parameters, do any of them want exposing to be able to be modified?

SPPearce commented on modules/nf-core/soupx/templates/soupx.R:
```diff
@@ -0,0 +1,58 @@
+sc <- SoupChannel(counts_raw, counts_filt, calcSoupProfile = FALSE)
+soup_prof <- data.frame(
+    row.names = rownames(counts_filt),
+    est       = rowSums(counts_filt) / sum(counts_filt),
```
This will fail if there are no counts.

LeonHafner committed:
Update modules/nf-core/soupx/tests/main.nf.test

Co-authored-by: Simon Pearce <24893913+SPPearce@users.noreply.github.com>

LeonHafner committed:
Update modules/nf-core/soupx/tests/main.nf.test

Co-authored-by: Simon Pearce <24893913+SPPearce@users.noreply.github.com>

LeonHafner committed:
Update modules/nf-core/soupx/main.nf

Co-authored-by: Simon Pearce <24893913+SPPearce@users.noreply.github.com>

LeonHafner committed:
Update modules/nf-core/soupx/main.nf

Co-authored-by: Simon Pearce <24893913+SPPearce@users.noreply.github.com>

LeonHafner committed:
add filename check

LeonHafner committed:
expose params and add test

LeonHafner committed:
check for counts

LeonHafner commented:
Added the following things:
                                                          
  1. Default prefix — `prefix = task.ext.prefix ?: "${meta.id}_soupx"` in both `script:` and `stub:`.                                                                                                                                                                          
  2. Input/output name disambiguation — errors out if `${prefix}_soupx.h5ad` collides with either input filename.
  3. `sanitizeOutput` — both snapshot assertions now use `snapshot(sanitizeOutput(process.out)).match()`.                                                                                                                                                                      
  4. Zero-counts guard — early `stop()` in the R template if the filtered counts matrix is empty, instead of failing cryptically inside SoupX.                                                                                                                                 
  5. Exposed parameters — `ext.npcs` (default 50) and `ext.cluster_algorithm` (default `leiden`, one of `louvain|louvain_multilevel|slm|leiden`). Documented in `meta.yml`; added a third nf-test case covering non-default args

LeonHafner committed:
Merge branch 'master' into soupx

LeonHafner committed:
switch to ext.args

LeonHafner committed:
linting

nictru committed:
Replace ext.args with explicit npcs and cluster_algorithm inputs in SOUPX.

Expose Seurat PCA and clustering parameters as dedicated module inputs so pipelines can configure them directly without parsing ext.args.

nictru committed:
Include r-base in environment

nictru committed:
Add @nictru as maintainer

nictru committed:
Remove redundant input comments from SOUPX main.nf.

Document npcs and cluster_algorithm in meta.yml only, matching common nf-core module style.

nictru committed:
Merge branch 'master' into soupx
